src/Google.Api.Gax/Platform.cs

```diff
@@ -167,56 +173,238 @@ public GaePlatformDetails(string gcloudProject, string gaeInstance, string gaeSe
+                    podJson = await client.GetStringAsync(urlPod).ConfigureAwait(false);
+                }
+                catch
+                {
+                    // Ignore all errors
+                }
+            }
+            return new KubernetesData
+            {
+                PodName = podName,
+                NamespaceJson = namespaceJson,
+                PodJson = podJson,
+                MountInfo = mountInfo
+            };
+        }
+#elif NET45
+        internal static Task<KubernetesData> LoadKubernetesDataAsync()
+        {
+            // .NET45 code cannot currently run on Kubernetes.
+            // And .NET45 does not support the certificate functionality that we require.
+            return Task.FromResult<KubernetesData>(null);
+        }
+#else
+#error Unsupported platform
+#endif
+
         /// <summary>
         /// Builds a <see cref="GkePlatformDetails"/> from the given metadata.
         /// This metadata is normally retrieved from the GCE metadata server.
         /// </summary>
         /// <param name="metadataJson">JSON metadata, normally retrieved from the GCE metadata server.</param>
         /// <returns>A populated <see cref="GkePlatformDetails"/> if the metadata represents and GKE instance;
         /// <c>null</c> otherwise.</returns>
-        public static GkePlatformDetails TryLoad(string metadataJson)
+        [Obsolete("Use TryLoad(string, KubernetesData) instead.")]
+        public static GkePlatformDetails TryLoad(string metadataJson) =>
+            TryLoad(metadataJson, GkePlatformDetails.LoadKubernetesDataAsync().Result);
+
+        /// <summary>
+        /// Builds a <see cref="GkePlatformDetails"/> from the given metadata and kubernetes data.
+        /// The metadata is normally retrieved from the GCE metadata server.
+        /// The kubernetes data is normally retrieved using the kubernetes API.
+        /// </summary>
+        /// <param name="metadataJson">JSON metadata, normally retrieved from the GCE metadata server.
+        /// Must not be <c>null</c>.</param>
+        /// <param name="kubernetesData">Kubernetes data, normally retrieved using the kubernetes API.
+        /// Must not be <c>null</c>.</param>
+        /// <returns>A populated <see cref="GkePlatformDetails"/> if the metadata represents and GKE instance;
+        /// <c>null</c> otherwise.</returns>
+        public static GkePlatformDetails TryLoad(string metadataJson, KubernetesData kubernetesData)
         {
-            JObject metadata;
+            GaxPreconditions.CheckNotNull(metadataJson, nameof(metadataJson));
+            GaxPreconditions.CheckNotNull(kubernetesData, nameof(kubernetesData));
+            JObject metadata = null;
+            JObject namespaceData = null;
+            JObject podData = null;
             try
             {
                 metadata = JObject.Parse(metadataJson);
+                namespaceData = JObject.Parse(kubernetesData.NamespaceJson);
```

Comment:
What happens if PodJson is null? Does JObject.Parse just return null? (I'm trying to think about all the possible partial-information scenarios.)

Reply:
Refactored slightly. It now doesn't matter if Parse returns null or throws, the behaviour will be the same.